controller/src/main/java/io/pravega/controller/store/stream/ZKStoreHelper.java

```diff
@@ -53,42 +52,46 @@ public ZKStoreHelper(final CuratorFramework cf, Executor executor) {
-            } catch (KeeperException.NodeExistsException e) {
-                throw StoreException.create(StoreException.Type.NODE_EXISTS, path);
-            } catch (Exception e) {
-                throw StoreException.create(StoreException.Type.UNKNOWN, path);
-            }
-        });
+        final CompletableFuture<Void> result = new CompletableFuture<>();
+        try {
+            client.create().creatingParentsIfNeeded().inBackground(
+                    callback(x -> result.complete(null),
+                            e -> {
+                                if (e instanceof DataExistsException) {
+                                    result.completeExceptionally(
+                                            StoreException.create(StoreException.Type.NODE_EXISTS, path));
+                                } else {
+                                    result.completeExceptionally(
+                                            StoreException.create(StoreException.Type.UNKNOWN, path));
+                                }
+                            }), executor).forPath(path);
+        } catch (Exception e) {
+            result.completeExceptionally(StoreException.create(StoreException.Type.UNKNOWN, path));
+        }
+        return result;
     }
 
     CompletableFuture<Void> deleteNode(final String path) {
-        return CompletableFuture.runAsync(() -> {
-            try {
-                client.delete().forPath(path);
-            } catch (KeeperException.NoNodeException e) {
-                throw StoreException.create(StoreException.Type.NODE_NOT_FOUND, path);
-            } catch (KeeperException.NotEmptyException e) {
-                throw StoreException.create(StoreException.Type.NODE_NOT_EMPTY, path);
-            } catch (Exception e) {
-                throw StoreException.create(StoreException.Type.UNKNOWN, path);
-            }
-        });
-    }
-
-    CompletableFuture<List<String>> getStreamsInPath(final String path) {
-        return CompletableFuture.supplyAsync(() -> {
-            try {
-                return client.getChildren().forPath(path);
-            } catch (KeeperException.NoNodeException e) {
-                FutureHelpers.failedFuture(StoreException.create(StoreException.Type.NODE_NOT_FOUND, path));
-            } catch (Exception e) {
-                FutureHelpers.failedFuture(StoreException.create(StoreException.Type.UNKNOWN, path));
-            }
-            return null;
-        });
+        final CompletableFuture<Void> result = new CompletableFuture<>();
+        try {
+            client.delete().inBackground(
+                    callback(x -> result.complete(null),
+                            e -> {
+                                if (e instanceof DataExistsException) {
+                                    result.completeExceptionally(
+                                            StoreException.create(StoreException.Type.NODE_NOT_EMPTY, path));
```

Comment:
Yes this means the node has children. This is a translation from Zookeeper's NotEmpty exception type(KeeperException.Code.NOTEMPTY). In PR https://github.com/pravega/pravega/pull/1382 we have renamed this to DATA_NOT_EMPTY. I'll carry over your comment to that PR, so we can think of a better name there.